### DISCUSSION TITLE
Redirect ego_graph from Dijkstra to Bellman Ford in case of negative weights

### DISCUSSION THREAD

**Discussion by LaurentBerder** [networkx/networkx#5082] [Ideas]
Hi,

I'm using OSMnx to create a graph and calculate all possible routes using a cost function (instead of a distance) and a cutoff value.

The _ego_graph_ function is perfect for this and works very well for me.

Except that in some cases, my cost can be negative, and that breaks the __dijkstra_multisource_ function that _ego_graph_ refers to since it doesn't accept any negative weights.

What I'm imagining is a way to first check the list of weight values in the graph to make sure that there aren't any negative values, and if there are, call an equivalent function to __dijkstra_multisource_ that accepts negative weights (probably a Bellman Ford function, but I haven't been able to find one that allows me to use a cutoff value).

You might want to make that optional (say if you know that all your weights are positive, you'd specify it and therefore save computational time).

That way, you'd be avoiding an error when it should (theoretically) be possible to reach a result.

Here's the code I'm using
```python
subgraph = nx.ego_graph(G,
                        n=closest_point,
                        distance='cost', radius=cutoff_value)
```

And I'd be looking for something similar with an additional parameter:
```python
subgraph = nx.ego_graph(G,
                        n=closest_point,
                        distance='cost', radius=cutoff_value,
                        negative_weights=True)
```

**Comment by dschult** [comment]
My understanding is that bellman-ford has trouble with a cutoff due to the algorithm itself.  Indeed, one problem with negative edge weights is that the last edge you look at might be very negative and thus you have to look at all edges.

How would you like the negative edge weights to be interpreted?  How can a OSM have a negative distance?

I suspect these are very slight negative values that might be finessed in another way.  For example, maybe you can set those edge weights to zero.  That would be quite appropriate if you truly believe that the distances aren't negative but you are dealing with round-off error or other approximations to real distances. 

Or you can always use Johnson or Bellman-Ford and then find the nodes within your desired radius.

**Reply by LaurentBerder** [reply]
Thanks for coming back, @dschult.

The weights I'm using are not distances, but energy costs. And on a downhill road for exemple, electric vehicles recover energy, so costs have to be negative in these cases.

That's why I was looking into Bellman-Ford, as I don't see a suitable way to render them positive and use Dijkstra.

> Or you can always use Johnson or Bellman-Ford and then find the nodes within your desired radius.

The radius of the available area with a designated amount of energy in the battery is in fact what I'm looking for (that's what I wanted a cutoff for), so I don't see how to look for nodes within the radius. Or maybe I'm looking at it through a wrong angle.

**Comment by dschult** [comment]
Understood...   It sounds like you do need to find the length to all other nodes (or at least explore all the edges in case one is very downhill... and at that point you might as well keep track of all shortest path lengths :)

If you have to do this with all the nodes, you could consider `all_pairs_bellman_ford_path_length`, but for a single node maybe something like this would work:
```python
distance_iterator = nx.single_source_bellman_ford_path_length(G, closest_point, weight="cost")
ego_nodes = (n for n, dist in distance_iterator if dist <= cutoff_value)
ego_graph = G.subgraph(ego_nodes).copy()
```

**Reply by LaurentBerder** [reply]
Using `all_pairs_bellman_ford_path_length` returns an iterator, and if I read a dict from it like the following

`t = dict(nx.all_pairs_bellman_ford_path_length(G, weight='cost'))`

I get this error:
```python
Traceback (most recent call last):
  File "venv\lib\site-packages\IPython\core\interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-31-2a87f00fc4e8>", line 1, in <module>
    t = dict(nx.all_pairs_bellman_ford_path_length(G, weight='cost'))
  File "venv\lib\site-packages\networkx\algorithms\shortest_paths\weighted.py", line 1663, in all_pairs_bellman_ford_path_length
    yield (n, dict(length(G, n, weight=weight)))
  File "venv\lib\site-packages\networkx\algorithms\shortest_paths\weighted.py", line 1541, in single_source_bellman_ford_path_length
    return _bellman_ford(G, [source], weight)
  File "venv\lib\site-packages\networkx\algorithms\shortest_paths\weighted.py", line 1307, in _bellman_ford
    raise nx.NetworkXUnbounded("Negative cost cycle detected.")
networkx.exception.NetworkXUnbounded: Negative cost cycle detected.
```

Using your proposed code 
```python
distance_iterator = nx.single_source_bellman_ford_path_length(G, closest_point, weight="cost")
ego_nodes = (n for n, dist in distance_iterator if dist <= cutoff_value)
ego_graph = G.subgraph(ego_nodes).copy()
```
I get the same error as above, telling me `Negative cost cycles detected`

Why would Bellman Ford refuse negative costs?

**Reply by dschult** [reply]
Bellman Ford isn't rejecting negative weights. It is rejecting a negative **cycle**.  If you can get a loop that has a negative total weight, then the shortest path has weight negative infinity if the path can reach that cycle. It just keeps going around getting shorter and shorter.  So there isn't a way to get shortest path if there is a negative cycle.

If your application is showing the car charging the battery around a loop you've discovered a perpetual motion machine. Could be really helpful with climate change. :}

**Reply by LaurentBerder** [reply]
Yeah, I'd better hurry up and put a patent on that! ;-)

I see, thanks.
That's strange indeed, because my graph being directed, there shouldn't be a perpetual loop: one way should be positive, the other should be negative. Right?

**Comment by dschult** [comment]
Definitely worth finding that negative cost cycle.   This example makes me realize that the error message might be able ti figure out what that cycle is -- but I seem to recall that the code for doing that was tricky. If you can figure out a way to have the code get the negative cycle rather than just reporting that the cycle exists, we could improve that error message quite a bit.

**Reply by LaurentBerder** [reply]
I'm starting to move into another direction, albeit a very slow one...
What I'm resorting to is generating all shortest paths on the network (that does take a lot of time) with `osmx.shortest_path(G, s, n, weight='length')`.
Then, I use the index of the edges of the shortest paths to look for the edges' weights in the graph.
And I look at the cumsum of these weights to find the points where I meet the threshold.

It's all a slow process, but it gets to the point.
Ideally I would love it if there was another equivalent solution that gave me the same results.

**Reply by dschult** [reply]
So, is it solving for shortest paths in terms of "length"?  And then looking up the total "cost" of the path?
That's not the same as what you were looking for (and it can be done in networkx with `nx.shortest_path(G, s, target=n, weight='length')` followed by computing the path "cost" summing the edges.

```python
path = nx.shortest_path(G, s, n, "length")
path_cost = sum(G[u][v]["cost"] for u, v in nx.utils.pairwise(path))
```

**Reply by LaurentBerder** [reply]
You're right, it's not exactly the same.

Your method looks quite quicker than what I'm doing.
However the second line doesn't work.
```python
Traceback (most recent call last):
  File "venv\lib\site-packages\IPython\core\interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-175-24d4a9a9f360>", line 1, in <module>
    sum(G[u][v]["cost"] for u, v in nx.utils.pairwise(path))
  File "<ipython-input-175-24d4a9a9f360>", line 1, in <genexpr>
    sum(G[u][v]["cost"] for u, v in nx.utils.pairwise(path))
  File "venv\lib\site-packages\networkx\classes\coreviews.py", line 51, in __getitem__
    return self._atlas[key]
KeyError: 'cost'
```
I thought it might have been a typo, and changed the line to 
```python 
path_cost = sum(G.edges[u, v, 0]["cost"] for u, v in nx.utils.pairwise(path))
```
However that still doesn't fix it:

```python
Traceback (most recent call last):
  File "venv\lib\site-packages\IPython\core\interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-176-2206664bae0c>", line 1, in <module>
    path_cost = sum(G.edges[u, v, 0]["cost"] for u, v in nx.utils.pairwise(path))
  File "<ipython-input-176-2206664bae0c>", line 1, in <genexpr>
    path_cost = sum(G.edges[u, v, 0]["cost"] for u, v in nx.utils.pairwise(path))
  File "venv\lib\site-packages\networkx\classes\reportviews.py", line 1201, in __getitem__
    return self._adjdict[u][v][k]
KeyError: 1861881342
```

I think what happens is that, because I didn't specify a destination in the first line (`path = nx.shortest_path(G, source=closest_point, weight="length")`), path (which is a dict of shape `{source: [node, node, destination]}` doesn't behave well with the `nx.utils.pairwise` function (it seems to couple `source` directly with `destination`, so the edges do not always exist.

If I do specify a destination in the first line (`path = nx.shortest_path(G, source=closest_point, target=destination, weight="length")`), the function works (only with the typo correction) but returns a single float (total cost for the whole path). This is not suitable for me, because I need to find the index of the node where `np.cumsum(cost) > threshold`.

**Reply by dschult** [reply]
Interesting -- I'll look into it more --- but first, what is your `G`? directed?  multigraph? And I'm assuming you have both attributes called `length` and also called `cost`.  Is that correct?

Also, I thought you were looking through all paths in the graph. Then you don't need to search within the path for the index where `cost > threshold` because you will be considering every node pair and so will look at these partial paths eventually anyway.

But I think the biggest problem with this approach is that you are finding the `cost` of the path of smallest `length`.  That is not the path of smallest `cost`.  But maybe Ia m missing something.

**Comment by LaurentBerder** [comment]
It's a graph that's created through OSMnx (which downloads Open Street Map data for the requested area), and it's a MultiDiGraph.

As I said above, you are indeed right in the sense that finding cost of path of smallest length and finding path of smallest cost are not the same.
However, first I need to reach some result at some point (and so far I don't see how to reach the latter even with Bellman Ford because of these negative cycles), and second I am quite sure the difference will not be too noticeable at the scale that I'm working on at the moment.

But if you do have a better idea that would allow me to get on the track of finding path of smallest cost, I'll be glad to explore it, implement it and share it with you.

**Comment by dschult** [comment]
The fact that G is a MultiGraph (also Di, but that doesn't affect this issue) means that more work has to be done to collect the shortest edges.  Basically, we have to find the minimum cost among all edges between each pair in the path:

```python
path = nx.shortest_path(G, source=s, target=n, "length")
path_cost = sum((min(datadict["cost"] for datadict in G[u][v].values()) for u, v in nx.utils.pairwise(path))
```
The code: `(min(datadict["cost"] for datadict in G[u][v].values())` finds the minimum cost of all edges between u and v.
The data structure is a dict-of-dict-of-dict-of-dict   keyed by   node, nbr, edgekey, edge attribute, to a value of the edge attribute.
You could also use `G.edges[u, v, k]["cost"]` and look over all `k` but this is closer to the data structure so I think ought to be faster.

I guess I would recommend spending some time figuring out where the negative cost cycles are coming from.  Is it round-off error, or bad data entry somewhere. It's hard for me to believe that any streetmap with energy costs on each edge would be even close to break-even let alone negative.  But unfortunately even 1 negative cycle breaks the whole algorithm. How best to fix it will depend on what is causing the negative cycles.

**Reply by LaurentBerder** [reply]
Well, since I'm looking for paths to all nodes (including some nodes that are very close to my origin node), if there are nodes that are lower in altitude than point of origin (and do not need to go through any higher point), it could be that the path is strictly negative, right?

To shorten my current code's execution time, I want to reduce the `nx.shortest_path` calls to the furthest nodes from my origin node. The origin point is at the center of a square of which I know the coordinates, so I was thinking about doing an intersection of a shapely Polygon with the graph, is that something achievable?
That woud actually also ensure that I don't have any negative cycles since, as you point out, for a sufficiently long path, there's no way the energy can be absolutely negative.

**Reply by dschult** [reply]
Negative paths are fine -- that won't cause any problems.  It is only negative cycles that cause trouble -- (perpetual motion machines).  So, its important to figure out how/why they exist. If it is just round-off errors then you can add 1e-10 to every weight and solve the problem. If it is a specific edge that seems to have bad data, then you can change that edge. Fixing bad data by changing how you solve the problem isn't always t he best approach.

**Reply by LaurentBerder** [reply]
Problem is I mut admit I have no idea where to start looking for that.

**Reply by rossbar** [reply]
`simple_cycles` might be helpful for this.

**Reply by dschult** [reply]
We ought to be able to find a negative cycle once we detect it (inside Bellman-ford). It might be time consuming in some cases, but we have a dict of predecessors and we know the cycle involves the currently considered node.  e..g inside `_bellman_ford` there is a place where the negative cycle is detected (consider heuristic==False case for starters). Just before raising the exception, could we loop back through the predecessors to find the cycle?

**Reply by LaurentBerder** [reply]
Right.

So on your recommandations I tried the following:
```python
cycles = nx.simple_cycles(G)
for cycle in cycles:
    path_length = nx.bellman_ford_path_length(G, source=closest_point, target=cycle, weight="cost")
```
And when it did break because of a `networkx.exception.NetworkXUnbounded: Negative cost cycle detected.`, I looked at the cycle. It's a 5913 long list of nodes.
That's pretty long, as my whole network only has 41093 nodes.

**Reply by dschult** [reply]
Wow -- that is a long list of nodes. But remember, that is the simple cycle.... not the negative cost cycle. You are finding a path from `closest_point` to *any* node in the target `cycle`. So, the fat that `cycle` is big, doesn't mean that the negative cost cycle found by `bellman_ford` is long.   How about looking for cycles in expanding neighborhoods:

```python
nbrs = [closest_point] + list(G[closest_point])
while len(nbrs) < len(G):
    Gsub = G.subgraph(nbrs).copy()
    path_length = nx.bellman_ford_path_length(Gsub, source=nbrs[-1], weight="cost")
    print(f"no negative cycle when considering neighborhood with {len(nbrs)} nodes")
    nbrset = {nbr for nbr in G[u] for u in Gsub}
    nbrset.update(Gsub)
    nbrs = list(nbrset)
```

**Reply by LaurentBerder** [reply]
```python
    path_length = nx.bellman_ford_path_length(Gsub, source=nbrs[-1], weight="cost")
TypeError: bellman_ford_path_length() missing 1 required positional argument: 'target'
```
So I corrected with what I think you mean:
```python
nbrs = [closest_point] + list(G[closest_point])
while len(nbrs) < len(G):
    Gsub = G.subgraph(nbrs).copy()
    path_length = nx.bellman_ford_path_length(Gsub, source=closest_point, target=nbrs[-1], weight="cost")
    print(f"no negative cycle when considering neighborhood with {len(nbrs)} nodes")
```
But then the next line doesn't work either.
It looks like you're trying to make a dict with the node_id as key, and the `G[node_id]` as value, am I correct?
In that case, it could be 
```python 
nbrset = {nbrs[i]: Gsub[nbrs[i]] for i in range(len(nbrs))}
```
Then again, I hit a roadbock on the following line, as `TypeError: cannot convert dictionary update sequence element #0 to a sequence`.

Again, I have to thank you for the time you take for helping me and for your reactivity!

**Reply by dschult** [reply]
Sorry...  I forgot the "target" argument to the function and then I switched the order of the set-comprehension.
It should be:  `nbrset = {nbr for u in Gsub for nbr in G[u]}`   and it builds a set with node elements that are neighbors in G of any node in Gsub.

But better than that (which just tells you how many edges the negative cycles have), take a look at #5089 where I am trying
to add a function that can find negative cycles.

**Reply by LaurentBerder** [reply]
Sorry, it took me a while to install `networkx-2.7rc1.dev0` for this to work.

So I ran the following
```python
neg_cycle = nx.find_negative_cycle(G, source=closest_point, weight='cost')
```
I then tried to analyse the result by looking at the costs of the edges for the negative cycle that was found (it returned a list with node ids, the last one being equal to the first one).
```python
[G.edges[(neg_cycle[i], neg_cycle[i+1], 0)].get('cost') for i in range(len(neg_cycle) - 1)]
```
And I got this:
```python
[0.015096362827270171,
 0.05437160427964326,
 -0.03060913199775348,
 -0.06502802565554765]
```
Indeed, the sum is negative at `-0.026169190546387694`.

I'll try to get rid of that cycle somehow.

**Reply by LaurentBerder** [reply]
Alright, I'm back after some interruption.

I got rid of my negative cycle, and am now ready to go through with Bellman Ford.
I tried the following:
```python
nx.weighted._bellman_ford(G, source=[closest_point], weight='cost')
```
But I'm getting an error
```python
Traceback (most recent call last):
  File "venv\lib\site-packages\IPython\core\interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-92-da6134434c32>", line 1, in <module>
    nx.weighted._bellman_ford(G, source=[closest_point], weight='cost')
  File "venv\lib\site-packages\networkx\algorithms\shortest_paths\weighted.py", line 1330, in _bellman_ford
    dist_v = dist_u + weight(u, v, e)
TypeError: 'str' object is not callable
```
I tried investigating, and I'm not sure how that line 1330 is supposed to work, because in `weight(u, v, e)`, both `u` and `v` are indeed `int`s, but `e` is a `networkx.classes.coreviews.AtlasView`.
But that's not the error I get. The error comes from the fact that my weight is the name of an attribute, not a function.

` `

So, I tried another function that you had suggested earlier, which wasn't working at the time because of the negative cycle.
```python
distance_iterator = nx.single_source_bellman_ford_path_length(G, closest_point, weight="cost")
ego_nodes = (n for n, dist in distance_iterator if dist <= cutoff_value)
ego_graph = G.subgraph(ego_nodes).copy()
```
```python
Traceback (most recent call last):
  File "venv\lib\site-packages\networkx\classes\graph.py", line 1911, in bunch_iter
    for n in nlist:
  File "<ipython-input-104-72b4fe5a9b62>", line 2, in <genexpr>
    ego_nodes = (n for n, dist in distance_iterator if dist <= cutoff_value)
TypeError: cannot unpack non-iterable int object
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "venv\lib\site-packages\IPython\core\interactiveshell.py", line 3418, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-104-72b4fe5a9b62>", line 3, in <module>
    ego_graph = G.subgraph(ego_nodes).copy()
  File "venv\lib\site-packages\networkx\classes\graph.py", line 1721, in subgraph
    induced_nodes = nx.filters.show_nodes(self.nbunch_iter(nodes))
  File "venv\lib\site-packages\networkx\classes\filters.py", line 52, in __init__
    self.nodes = set(nodes)
  File "venv\lib\site-packages\networkx\classes\graph.py", line 1926, in bunch_iter
    raise exc
networkx.exception.NetworkXError: nbunch is not a node or a sequence of nodes.
```

**Reply by dschult** [reply]
The function `_bellman_ford` is a private function that assumes the `weight` argument has already been processed to be a function. The processing consists of `weight = _weight_function(G, weight)` where `_weight_function` is another private function in the same `weighted.py` module.  It might be easier to use the public functions (the ones that don't have a `_` underscore in front of the name. For example:
```nx.weighted.bellman_ford_path_length(G, source=closest_point, weight='cost')```
They do all the argument processing needed to get the input ready for the private function which does most of the work.

**Reply by LaurentBerder** [reply]
I see, sorry, that calling an private function was a wrong way for me to proceed.

`nx.weighted.bellman_ford_path_length(G, source=closest_point, weight='cost')` doesn't work as it's missing the `target` argument.

What function should I be using to find the cumulative sum of costs for edges to all nodes from a specific start node?

**Reply by dschult** [reply]
That would be `nx.single_source_bellman_ford`.
See also:  
[the docs for all shortest path functions](https://networkx.org/documentation/stable/reference/algorithms/shortest_paths.html)

**Reply by LaurentBerder** [reply]
That's it!
Thanks a lot @dschult, that is indeed what I was looking for. It has required a lot of attention from you, sorry about that.

I still think it'd be a nice idea to be able to integrate bellman ford in ego_graph (as a possibility), but at least I've managed to get there manually.
